mlir/lib/Conversion/AIRToAIEPass.cpp

```diff
@@ -3061,16 +3234,65 @@ class AIRToAIEPass : public air::impl::AIRToAIEBase<AIRToAIEPass> {
 
         builder.setInsertionPointToEnd(device.getBody());
         auto keep_pkt_header = builder.getBoolAttr(true);
-        AIE::PacketFlowOp pktFlow = builder.create<AIE::PacketFlowOp>(
-            builder.getUnknownLoc(), flowID++, keep_pkt_header);
-        Region &r_pktFlow = pktFlow.getPorts();
-        Block *b_pktFlow = builder.createBlock(&r_pktFlow);
-        builder.setInsertionPointToStart(b_pktFlow);
-        builder.create<AIE::PacketSourceOp>(builder.getUnknownLoc(), srcTile,
-                                            AIE::WireBundle::Trace, 0);
-        builder.create<AIE::PacketDestOp>(builder.getUnknownLoc(), destTile,
-                                          AIE::WireBundle::DMA, destChan);
-        builder.create<AIE::EndOp>(builder.getUnknownLoc());
+        (void)createPacketFlowOp(
+            builder, flowID, srcTile, AIE::WireBundle::Trace, 0, destTile,
+            AIE::WireBundle::DMA, destChan, keep_pkt_header);
+      }
+    }
+  }
+
+  // Create an overlay of packet flow network sending control packets from shim
+  // dma mm2s channels to tiles in a column.
+  void createControlPacketFlow(AIE::DeviceOp device) {
```

Comment:
As far as I can tell this is a mlir-aie transformation. Can we put it there instead? If it was generalized a bit then it could already be used to fix [this](https://github.com/Xilinx/mlir-aie/blob/main/lib/Targets/AIETargetCDODirect.cpp#L611-L617) which is just a very small control overlay. It could be inserted with a pass instead. And I can think of several other uses.

Reply:
Yeah that makes great sense to me. It could totally become an mlir-aie pass, because it was only looking at `aie.tile` ops in the IR.

This overlay being generated comes with some constraints though. One thing that I haven't figure out is, if the mlir-aie design already has circuit-switch routed shim dma mm2s channel usage, then we certainly cannot use those channels for control packets. In AIR that was easily controlled, because we generate the whole design, and we can force the design to never circuit-switch reserve any shim s2mm channels to begin with. Whereas if this one becomes an mlir-aie pass, then I haven't figured out how to make this overlay flexible to an existing aie design, while still enfocing full connectivitly to every tile's control port...

Maybe we can discuss this in greater details in our meeting on Friday?